### PR TITLE
Refactor and combine STM32H7 linker scripts

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 

Up until now, the linker scripts for STM32H7 have been quite difficult to maintain, because there were 10+ different MCUs with their own linker scripts, and each of these linker scripts was slightly different from the others.  This caused a number of issues, such as:
- Many of the processor linker scripts were missing the special sections required for Ethernet support, so ethernet did not work reliably on those processors (the MPU regions wouldn't get set up right)
- STM32H7 devices are good candidates for split heap support, but only one MCU actually had that enabled in the linker script
- Crash capture memory was only set up for one MCU, not the rest of them

To fix those issues, and to make maintenance easier going forward, I combined all of those linker scripts together into just four scripts:
- One for the H743 and H72x families
- One for the H745/47 family CM7 core
- One for the H745/47 family CM4 core
- One for the H7Ax family

(refer to [here](https://github.com/mbed-ce/mbed-os/wiki/MCU-Info-Page:-STM32H7-Family) if you need a refresher on what the different families mean).

As part of this, I also reworked how memory is handled in the STM32H7 ethernet driver.  Instead of relying on rickety absolute addresses and fixed MPU regions (that also conflicted with the Mbed MPU code, oops), it now uses a dynamically sized MPU region that's set up the same in all of the linker scripts to handle the descriptors.  And for the buffers, those don't need an MPU region, so we can just stick those in normal memory and use CacheAlignedBuffer to handle the cache-DMA coherency.

To be honest, I am proud of this refactor and I would really like to start doing more linker scripts in this manner.  It is much cleaner and more resilient to config changes!

#### Impact of changes 
- Bug on some STM32H7 devices where heap allocations would fail since the memory bank info PR is fixed
- All STM32H7 MCUs now support ethernet, crash capture, and split heap in the linker scripts
- All STM32H7 linker scripts now use memory bank information from cmsis_mcu_descriptions.json, meaning they can adapt to new memory configurations easily
- STM32H7 ethernet MAC code updated to use new memory layout and be smarter about MPU configuration

### Documentation 

None

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [X] Tests / results supplied as part of this PR
    
Working on running tests...
----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H725xE/TARGET_WIO_H725AE/CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,6 @@ add_library(mbed-wio-h725ae INTERFACE)
 target_sources(mbed-wio-h725ae
     INTERFACE
         PeripheralPins.c
-        system_clock_override.c
```
Comment (multiplemonomials) [inline targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H725xE/TARGET_WIO_H725AE/CMakeLists.txt:9]: Why did this one still have a system clock file?  I thought I got rid of all those...

File: targets/TARGET_STM/TARGET_STM32H7/linker_scripts/stm32_eth_region_size_calcs.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * This file contains calculations for the size of the memory region used for the Ethernet buffer descriptors.
+ * This region must be a power-of-2 size so that it can be used as an MPU region.  Also, in dual core
+ * CPUs, the CM4 and CM7 have to agree on its size so they don't define conflicting memory regions.
+ * 
+ * So, this header has some fancy math to calculate it.  Note that this header is included by
+ * linker scripts so it can only contain preprocessor code.
+ */
+
+/* from here: https://stackoverflow.com/questions/22925016/rounding-up-to-powers-of-2-with-preprocessor-constants */
+#define POW2_CEIL(v) (1 + \
```
Comment (multiplemonomials) [inline targets/TARGET_STM/TARGET_STM32H7/linker_scripts/stm32_eth_region_size_calcs.h:28]: I know this function is really, really gross, but I really couldn't find another way to make this possible.  In order to be able to compute the MPU region size, we need a way to round up to the nearest power of 2.

File: targets/targets.json5 [hunk 3]
```diff
@@ -3481,47 +3476,24 @@
...
+            "enable-overdrive-mode": 0
         },
         "image_url": "https://www.st.com/bin/ecommerce/api/image.PF262651.en.feature-description-include-personalized-no-cpn-medium.jpg"
     },
-    "NUCLEO_H745ZI_Q" : {
-        "inherits" : [
-            "NUCLEO_H745ZI_Q_CM7"
-        ]
+    "NUCLEO_H745ZI_Q_CM7": {
+        inherits: ["MCU_STM32H745xI_CM7", "NUCLEO_H745ZI_Q"]
```
Comment (multiplemonomials) [inline targets/targets.json5:3493]: I decided it was time for a bit of a refactor here.  Mbed targets support multiple inheritance, so for boards that have dual core MCUs, we can have one non-public target for the board and another for each MCU core, and then combine them together like this.  It removes some copy-pasted code and is a bit cleaner.

File: targets/TARGET_STM/TARGET_STM32H7/linker_scripts/stm32_eth_region_size_calcs.h (hunk 2 above)
Comment (JohnK1987) [reply targets/TARGET_STM/TARGET_STM32H7/linker_scripts/stm32_eth_region_size_calcs.h:28]: Did you try to ask an AI for a optimization? :D I do not trust these tools but who know...

Comment (multiplemonomials) [reply targets/TARGET_STM/TARGET_STM32H7/linker_scripts/stm32_eth_region_size_calcs.h:28]: Honestly I trust the random SO post.  And I did spot check it and it seemed to be producing sane values.

**Comment by multiplemonomials** [comment]
I ran Mbed tests with this PR on STM32H743, STM32H723, and STM32H747, and did not see any notable new failures.  So I think it should be good to merge!

**Review by JohnK1987 — APPROVED** [review]
I am just confused about names of linker variants. We will see in the future how this will be good for orientation.

**Pull request merged**